docs/tex/gateway_kernel_init_sequence_diagram/gateway_kernel_init_sequence_diagram.tex

```diff
@@ -0,0 +1,43 @@
+\documentclass[convert]{standalone}
+\usepackage{fontspec}
+\usepackage{pgf-umlsd}
+\usepackage[bottom]{footmisc}
+
+\setmainfont{Consolas}
+
+\begin{document}
+\begin{sequencediagram}
+    \newthread{browser}{:browser}
+    \newinst[4]{server}{:jupyter\_server}
+    \newinst[3]{gateway}{:enterprise\_gateway}
+    \newinst[4]{kernel}{:kernel}
+    
+    \begin{sdblock}{fetch kernelspecs}{}
+        \begin{call}{browser}{https GET api/kernelspecs}{server}{api/kernelspecs response}
+            \begin{call}{server}{https GET api/kernelspecs}{gateway}{api/kernelspecs response}
+            \postlevel
+            \end{call}
+        \end{call}
+    \end{sdblock}
+    
+    \begin{sdblock}{kernel init}{}
+        \begin{call}{browser}{https POST api/sessions}{server}{api/sessions response}
+            \begin{call}{server}{https POST api/sessions}{gateway}{api/sessions response}
+                \begin{call}{gateway}{CoreV1API.create\_namespaced\_pod}{kernel}{kernel connection info\footnote{see https://github.com/jupyter/enterprise\_gateway/pull/877}}
```

Comment:
If we wanted to make this generic, we could have something like:
```suggestion
                \begin{call}{gateway}{RemoteProcessProxy.launch\_process}{kernel}{kernel connection info\footnote{see https://github.com/jupyter/enterprise\_gateway/pull/877}}
```
There's also a _discovery_ step performed by the process-proxy immediately following the response of the connection info.  This is what actually determines the IP to use for the zmq ports (in addition to ensuring the kernel is "visible" via the resource manager's API).

I suspect this would need to add a "bar" for the resource manager in order to show this.  Reverse engineering this, I suppose this would add something like...

```
       \newinst[5]{rm}{:resource manager}
...
       \begin{call}{gateway}{RemoteProcessProxy.confirm\_remote\_startup}{rm}{status and host info}
```

Reply:
> There's also a _discovery_ step performed by the process-proxy immediately following the response of the connection info. This is what actually determines the IP to use for the zmq ports (in addition to ensuring the kernel is "visible" via the resource manager's API).
> 
> I suspect this would need to add a "bar" for the resource manager in order to show this. Reverse engineering this, I suppose this would add something like...
> 
> ```
>        \newinst[5]{rm}{:resource manager}
> ...
>        \begin{call}{gateway}{RemoteProcessProxy.confirm\_remote\_startup}{rm}{status and host info}
> ```

@kevin you're going to have to explain this one a little better. Does information also end up flowing back the frontend from this, or just to the gateway? What's a resource manager in this context? For eg the k8s case, is the resource manager the k8s API itself?

Reply:
> Does information also end up flowing back the frontend from this, or just to the gateway? 

No. This phase is only between the remote-process-proxy (EG) and the RM.

> What's a resource manager in this context? For eg the k8s case, is the resource manager the k8s API itself?

Yes, for K8s it's the k8s API - and issued by the `KubernetesProcessProxy`.  Same thing for Hadoop YARN (via the `YarnClusterProcessProxy`).  We use a yarn python API that issues REST queries against the YARN RM.  Both of these return state and host information.  We then set the IP in the connection information from this host info.